pkg/security/api/v1/types.go

```diff
@@ -0,0 +1,86 @@
+	PodSpec kapi.PodSpec `json:"podSpec"`
+
+	// User is the user you're testing for.
+	// If you specify "User" but not "Group", then is it interpreted as "What if User were not a member of any groups.
+	// If User and Groups are empty, then the check is performed using *only* the ServiceAccountName in the PodSpec.
+	User string `json:"user,omitempty"`
+
+	// Groups is the groups you're testing for.
+	Groups []string `json:"groups,omitempty"`
+}
+
+// PodSecurityPolicySubjectReviewStatus contains information/status for PodSecurityPolicySubjectReview.
+type PodSecurityPolicySubjectReviewStatus struct {
+	// AllowedBy is a reference to the rule that allows the PodSpec.  A `nil`, indicates that it was denied.
+	AllowedBy *kapi.ObjectReference `json:"allowedBy"`
+
+	// Reason is optional.  It indicates why a request was allowed or denied.
+	Reason string `json:"reason,omitempty"`
+
+	// PodSpec is the PodSpec after the defaulting is applied.
+	PodSpec kapi.PodSpec `json:"podSpec"`
+}
+
+// PodSecurityPolicySelfSubjectReview checks whether this user/SA tuple can create the PodSpec
+type PodSecurityPolicySelfSubjectReview struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// Spec defines specification the PodSecurityPolicySelfSubjectReview.
+	Spec PodSecurityPolicySelfSubjectReviewSpec `json:"spec"`
+
+	// Status represents the current information/status for the PodSecurityPolicySelfSubjectReview.
+	Status PodSecurityPolicySubjectReviewStatus `json:"status,omitempty"`
+}
+
+// PodSecurityPolicySelfSubjectReviewSpec contains specification for PodSecurityPolicySelfSubjectReview.
+type PodSecurityPolicySelfSubjectReviewSpec struct {
+	// PodSpec is the PodSpec to check.
+	PodSpec kapi.PodSpec `json:"podSpec"`
+}
+
+// PodSecurityPolicyReview checks which service accounts (not users, since that would be cluster-wide) can create the `PodSpec` in question.
+type PodSecurityPolicyReview struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// Spec is the PodSecurityPolicy to check.
+	Spec PodSecurityPolicyReviewSpec `json:"spec"`
+
+	// Status represents the current information/status for the PodSecurityPolicyReview.
+	Status PodSecurityPolicyReviewStatus `json:"status,omitempty"`
+}
+
+// PodSecurityPolicyReviewSpec defines specification for PodSecurityPolicyReview
+type PodSecurityPolicyReviewSpec struct {
+	// PodSpec is the PodSpec to check. The ServiceAccountName field is ignored for this check.
+	PodSpec kapi.PodSpec `json:"spec"`
+	
+	// ServiceAccountNames is an optional set of ServiceAccounts to run the check with
+	ServiceAccountNames []string `json:"serviceAccountNames"`
+}
+
+// PodSecurityPolicyReviewStatus represents the status of PodSecurityPolicyReview.
+type PodSecurityPolicyReviewStatus struct {
+	// AllowedServiceAccounts returns the list of service accounts in *this* namespace that have the power to create the PodSpec.
+	AllowedServiceAccounts map[string]PodSecurityPolicySubjectReviewStatus `json:"allowedServiceAccounts"`
```

Comment:
@smarterclayton @deads2k what about a classic array of `(key,value)`? In this case we don't need O(1) access or grant uniqueness as David was mentioning it's only for the status.



```go
/ PodSecurityPolicyReviewStatus represents the status of PodSecurityPolicyReview.
type PodSecurityPolicyReviewStatus struct {
	// AllowedServiceAccounts returns the list of service accounts in *this* namespace that have the power to create the PodSpec.
	AllowedServiceAccounts []AllowedServiceAccountPair
}

// AllowedServiceAccountPair represents a pair of ServiceAccount name and related review status
type AllowedServiceAccountPair struct {
    // name contains the allowed ServiceAccount
    Name string
    
    // reviewStatus contains the review status for the ServiceAccount
    ReviewStatus PodSecurityPolicySubjectReviewStatus
}
```
Don't hesitate to rephrase or rename :)